src/extra/Markdown/index.tsx

```diff
@@ -138,11 +139,110 @@ type components = {
 	[element: string]: (props: any) => JSX.Element | React.ReactElement | Element;
 };
 
+interface Decorator {
+	match: RegExp;
+	component: string;
+}
+
+interface DecoratorPluginOptions {
+	decorators: Decorator[];
+}
+
+const decoratorPlugin = (options: DecoratorPluginOptions) => {
+	options = options || { decorators: [] };
+
+	if (!options.decorators || !options.decorators.length) {
+		return;
+	}
+
+	const targets: Array<{ decorator: Decorator; node: any }> = [];
+
+	const visitor = (node: any) => {
+		const matchedDecorator = options.decorators.find((decorator) => {
+			return (
+				node.children &&
+				node.children[0] &&
+				node.children[0].type === 'text' &&
+				decorator.match.test(node.children[0].value)
+			);
+		});
+
+		if (matchedDecorator) {
+			targets.push({
+				decorator: matchedDecorator,
+				node,
+			});
+		}
+	};
+
+	const modify = () => {
+		targets.forEach((target) => {
+			const node = target.node;
+			const match = target.decorator.match;
+
+			const text = node.children[0].value;
+			const children = [];
+
+			match.lastIndex = 0;
+			do {
```

Comment:
Let's try to refactor this whole function since it's quite difficult to follow what's going on, I couldn't even understand it after reading it for 5 minutes. Let's avoid using `do while` and clean up what is pushed to the children, how we store indices, etc.